driver-core/src/main/java/com/datastax/driver/core/NettyUtil.java

```diff
@@ -0,0 +1,156 @@
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import java.lang.reflect.Constructor;
+import java.util.Locale;
+import java.util.concurrent.ThreadFactory;
+
+import com.google.common.base.Throwables;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A set of utilities related to the underlying Netty layer.
+ */
+@SuppressWarnings("unchecked")
+class NettyUtil {
+
+    private static final boolean FORCE_NIO = SystemProperties.getBoolean("com.datastax.driver.FORCE_NIO", false);
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyUtil.class);
+
+    private static final boolean SHADED;
+
+    private static final boolean USE_EPOLL;
+
+    private static final Constructor<? extends EventLoopGroup> EPOLL_EVENT_LOOP_GROUP_CONSTRUCTOR;
+
+    private static final Class<? extends SocketChannel> EPOLL_CHANNEL_CLASS;
+
+    private static final Class[] EVENT_GROUP_ARGUMENTS = { int.class, ThreadFactory.class };
+
+    static {
+        boolean shaded;
+        try {
+            // prevent this string from being shaded
+            Class.forName(String.format("%s.%s.channel.Channel", "io", "netty"));
+            shaded = false;
+        } catch (ClassNotFoundException e) {
+            try {
+                Class.forName("com.datastax.shaded.netty.channel.Channel");
+                shaded = true;
+            } catch (ClassNotFoundException e1) {
+                throw new AssertionError("Cannot locate Netty classes in the classpath:" + e1);
+            }
+        }
+        SHADED = shaded;
+        boolean useEpoll = false;
+        try {
+            Class<?> epool = Class.forName("io.netty.channel.epoll.Epoll");
+            if (FORCE_NIO) {
+                LOGGER.info("Found Netty's native epoll transport in the classpath, "
+                    + "but NIO was forced through the FORCE_NIO system property.");
+            } else if (!System.getProperty("os.name", "").toLowerCase(Locale.US).equals("linux")) {
+                LOGGER.warn("Found Netty's native epoll transport, but not running on linux-based operating " +
+                    "system. Using NIO instead.");
+            } else if (!(Boolean)epool.getMethod("isAvailable").invoke(null)) {
```

Comment:
Indeed, but the reason why I kept the `os.name` check is that with `Epoll.unavailabilityCause()` you would have to print the stack trace to know why it didn't work. With the `os.name` check you would have this:
```
    0      [main] WARN  com.datastax.driver.core.NettyUtil  - Found Netty's native epoll transport in the classpath, but epoll is not available. Using NIO instead.
java.lang.ExceptionInInitializerError
	at io.netty.channel.epoll.Epoll.<clinit>(Epoll.java:30)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:190)
	at com.datastax.driver.core.NettyUtil.<clinit>(NettyUtil.java:68)
	at com.datastax.driver.core.NettyOptions.eventLoopGroup(NettyOptions.java:98)
	at com.datastax.driver.core.NettyOptions$$EnhancerByMockitoWithCGLIB$$7000decb.CGLIB$eventLoopGroup$1(<generated>)
	at com.datastax.driver.core.NettyOptions$$EnhancerByMockitoWithCGLIB$$7000decb$$FastClassByMockitoWithCGLIB$$a8371f19.invoke(<generated>)
	at org.mockito.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:216)
	at org.mockito.internal.creation.AbstractMockitoMethodProxy.invokeSuper(AbstractMockitoMethodProxy.java:10)
	at org.mockito.internal.invocation.realmethod.CGLIBProxyRealMethod.invoke(CGLIBProxyRealMethod.java:22)
	at org.mockito.internal.invocation.realmethod.FilteredCGLIBProxyRealMethod.invoke(FilteredCGLIBProxyRealMethod.java:27)
	at org.mockito.internal.invocation.InvocationImpl.callRealMethod(InvocationImpl.java:112)
	at org.mockito.internal.stubbing.answers.CallsRealMethods.answer(CallsRealMethods.java:36)
	at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:93)
	at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:29)
	at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:38)
	at org.mockito.internal.creation.MethodInterceptorFilter.intercept(MethodInterceptorFilter.java:61)
	at com.datastax.driver.core.NettyOptions$$EnhancerByMockitoWithCGLIB$$7000decb.eventLoopGroup(<generated>)
	at com.datastax.driver.core.NettyOptionsTest.should_invoke_netty_options_hooks(NettyOptionsTest.java:56)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: java.lang.IllegalStateException: Only supported on Linux
	at io.netty.channel.epoll.Native.<clinit>(Native.java:47)
	... 49 more
```